File: Sources/Kernel/MM/PMM.swift
/*
 * MM/PMM.swift
 * Physical Memory Manager
 */

import CSupport

public struct PhysAddr: Equatable {
    public var value: UInt64
    public init(_ value: UInt64) { self.value = value }
}

nonisolated(unsafe)  // Global state to avoid class initialization issues
    private var nextFree: UInt64 = 0x0800_0000
private let ramEnd: UInt64 = 0x2000_0000  // 512MB

public struct PMM {
    // Primitive Bump allocator for now.
    public static func allocateFrame() -> PhysAddr? {
        if nextFree >= ramEnd { return nil }
        let frame = nextFree
        nextFree += 4096
        // Clear frame content
        if let ptr = UnsafeMutableRawPointer(bitPattern: UInt(frame)) {
            memset(ptr, 0, 4096)
        }
        return PhysAddr(frame)
    }

    public static func allocateFrames(count: Int) -> PhysAddr? {
        let size = UInt64(count * 4096)
        if nextFree + size >= ramEnd { return nil }
        let frame = nextFree
        nextFree += size
        // Clear frame content
        if let ptr = UnsafeMutableRawPointer(bitPattern: UInt(frame)) {
            memset(ptr, 0, Int(size))
        }
        return PhysAddr(frame)
    }

    public static func freeFrame(_ frame: PhysAddr) {
        // Leak
    }
}

File: zig-kernel/src/pmm.zig
// Simple Physical Memory Manager (bump allocator)
// Ported from kernel/Sources/Kernel/MM/PMM.swift

extern fn memset(dest: [*]u8, c: u8, n: usize) [*]u8;

pub var nextFree: usize = 0x0800_0000; // start of free physical memory
const ramEnd: usize = 0x2000_0000; // 512MB

pub fn allocateFrame() ?usize {
    if (nextFree >= ramEnd) return null;
    const frame = nextFree;
    nextFree += 4096;
    const ptr: [*]u8 = @ptrFromInt(frame);
    _ = memset(ptr, 0, 4096);
    return frame;
}

pub fn allocateFrames(count: usize) ?usize {
    const size = count * 4096;
    if (nextFree + size >= ramEnd) return null;
    const frame = nextFree;
    nextFree += size;
    const ptr: [*]u8 = @ptrFromInt(frame);
    _ = memset(ptr, 0, size);
    return frame;
}

pub fn freeFrame(frame: usize) void {
    _ = frame;
    // No-op for now (leak)
}
